src/sentry/api/authentication.py

```diff
@@ -298,6 +299,39 @@ def authenticate(self, request: Request):
 class UserAuthTokenAuthentication(StandardAuthentication):
     token_name = b"bearer"
 
+    def _find_or_update_token_by_hash(self, token_str: str) -> ApiToken:
+        """
+        Find token by hash or update token's hash value if only found via plaintext.
+        1. Hash provided plaintext token.
+        2. Perform lookup based on hashed value.
+        3. If found, return the token.
+        4. If not found, search for the token based on its plaintext value.
+        5. If found, update the token's hashed value and return the token.
+        6. If not found via hash or plaintext value, raise AuthenticationFailed
+        """
+
+        hashed_token = hashlib.sha256(token_str.encode()).hexdigest()
+
+        try:
+            # Try to find the token by its hashed value first
+            return ApiToken.objects.select_related("user", "application").get(
+                hashed_token=hashed_token
+            )
+        except ApiToken.DoesNotExist:
+            try:
+                # If we can't find it by hash, use the plaintext string
+                api_token = ApiToken.objects.select_related("user", "application").get(
+                    token=token_str
+                )
+            except ApiToken.DoesNotExist:
+                # If the token does not exist by plaintext either, it is not a valid token
+                raise AuthenticationFailed("Invalid token")
+            else:
+                # Update it with the hashed value if found by plaintext
+                api_token.hashed_token = hashed_token
+                api_token.save(update_fields=["hashed_token"])
```

Comment:
This save() will generate an outbox to update the apitokenreplica table. Backfilling existing tokens will be hard because django migrations don't have access to runtime models. Instead migrations only have synthetic models that django generates.

We also have tooling to trigger a replication backfill via outboxes and options for replicated tables. Once your backfill on `apitoken` is complete, we could use the outbox backfill to sync the replica table. We have done a few of these backfills in the past.

Doing an outbox replica backfill requires defining an option (`outbox_replication.apitoken.replication_version`) and setting a `replication_version` on the source model (ApiToken). Once the option value is higher than the model version, the application will start processing the backfill.

Reply:
Adding `hashed_token` to `ApiTokenReplica` in this PR: https://github.com/getsentry/sentry/pull/66679